### PR TITLE
[X86] Don't save/restore fp around longjmp instructions

### PR DESCRIPTION
**Pull request by weiguozhi**
Longjmp instructions can also modify fp register, it is expected behavior. We should not save/restore fp around these instructions.

### CODE REVIEW

File: llvm/lib/Target/X86/X86MachineFunctionInfo.h [hunk 1]
```diff
@@ -174,6 +175,9 @@ class X86MachineFunctionInfo : public MachineFunctionInfo {
   bool FPClobberedByCall = false;
   bool BPClobberedByCall = false;
 
+  // A set of instructions clobber FP generated from longjmp.
+  SmallSet<MachineInstr *, 2> LongJmpMIClobberFP;
```
Comment (rnk) [inline llvm/lib/Target/X86/X86MachineFunctionInfo.h:179]: I don't think it is safe to store MIR pointers in side tables from pass-to-pass. It is too likely that a MachineInstr will be cloned and replaced, invalidating these pointers, leading to an ABA-style use-after-free.

It also occurs to me that we may need to make sure that X86MachineFunctionInfo is fully serialized through MIR tests. I would have to look further into that.

Comment (weiguozhi) [reply llvm/lib/Target/X86/X86MachineFunctionInfo.h:179]: Longjmp modifies fp and sp, these instructions can be treated as destroying current stack frame and setting up stack frame for the jump target.

How about set FrameDestroy flag for these longjmp instructions, then spillFPBP can ignore these instructions.

Comment (rnk) [reply llvm/lib/Target/X86/X86MachineFunctionInfo.h:179]: I like that approach, although there is some risk that the FrameDestroy flag is used to detect epilogue code. Longjmp instructions are effectively epilogue code: it tears down the frame and replaces it.

**Review by rnk — APPROVED** [review]
Thanks!

**Comment by GameCubeGBA** [comment]
Should this be backported to 19.x?

**Comment by weiguozhi** [comment]
> Should this be backported to 19.x?

This is a bug fix for https://github.com/llvm/llvm-project/commit/0d471b3f64d3116bd57c79d872f7384fff80daa5. So it depends on whether https://github.com/llvm/llvm-project/commit/0d471b3f64d3116bd57c79d872f7384fff80daa5 is included in 19.x.

**Pull request merged**